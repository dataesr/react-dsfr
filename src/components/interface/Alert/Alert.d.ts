import * as React from 'react';

export type AlertAs = 'p' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export type AlertTitle = string | Object | any[];

export type AlertDescription = string | Object | any[];

export type AlertType = 'error' | 'success' | 'info';

export type AlertClassName = string | Object | any[];

export interface AlertProps {
  as?: AlertAs;
  title: AlertTitle;
  description?: AlertDescription;
  type?: AlertType;
  small?: boolean;
  role?: string;
  show?: boolean;
  closable?: boolean;
  onClose?: (...args: any[])=>any;
  className?: AlertClassName;
}

declare const Alert: React.FC<AlertProps>;

export default Alert;
