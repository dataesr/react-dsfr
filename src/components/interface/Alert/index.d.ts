import type Alert from './Alert';

export type { Alert };
