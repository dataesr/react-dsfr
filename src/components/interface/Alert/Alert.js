import React, { useEffect, useState } from 'react';

import PropTypes from 'prop-types';
import classNames from 'classnames';

import dataAttributes from '../../../utils/data-attributes';

import '@gouvfr/dsfr/dist/component/alert/alert.css';

/**
 *
 * @visibleName Alert
 */
const Alert = ({
  className,
  as,
  type,
  title,
  description,
  role,
  small,
  show,
  closable,
  onClose,
  ...remainingProps
}) => {
  const HtmlTitleTag = `${as}`;
  const [internalShow, setInternalShow] = useState(show);
  const _className = classNames(
    'fr-alert',
    `fr-alert--${type}`,
    {
      'fr-alert--sm': small,
    },
    className,
  );

  useEffect(() => {
    setInternalShow(show);
  }, [show]);

  if (!internalShow) {
    return null;
  }

  return (
    <div
      role={role || undefined}
      className={_className}
      {...dataAttributes.getAll(remainingProps)}
    >
      <HtmlTitleTag className="fr-alert__title">{title}</HtmlTitleTag>
      {!small && description && <p>{description}</p>}
      {closable && (
        <button
          type="button"
          className="fr-link--close fr-link"
          onClick={() => {
            setInternalShow(false);
            if (onClose) {
              onClose();
            }
          }}
        >
          Masquer le message
        </button>
      )}
    </div>
  );
};

Alert.defaultProps = {
  as: 'p',
  type: 'info',
  role: '',
  description: undefined,
  small: false,
  show: true,
  closable: false,
  onClose: undefined,
  className: '',
};

Alert.propTypes = {
  as: PropTypes.oneOf(['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]).isRequired,
  description: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  type: PropTypes.oneOf(['error', 'success', 'info']),
  small: PropTypes.bool,
  role: PropTypes.string,
  show: PropTypes.bool,
  closable: PropTypes.bool,
  onClose: PropTypes.func,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
};

export default Alert;
