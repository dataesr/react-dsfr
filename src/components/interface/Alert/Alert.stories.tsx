import { ComponentStory } from '@storybook/react';

import Alert from './Alert';

export default {
  title: 'Alert',
  component: Alert,
};

const Template: ComponentStory<typeof Alert> = (args) => (
  <Alert {...args} />
);

export const Information = Template.bind({});
Information.args = {
  title: 'Information',
  description: "Description de l'alerte",
};
Information.storyName = 'Information';

export const Succes = Template.bind({});
Succes.args = {
  title: 'Succes',
  description: 'Vous avez reussi',
  type: 'success',
};
Succes.storyName = 'Succes';

export const Erreur = Template.bind({});
Erreur.args = {
  title: 'Erreur',
  description: "Impossible d'effectuer cette action",
  type: 'error',
};
Erreur.storyName = 'Erreur';

export const Petite = Template.bind({});
Petite.args = {
  title: 'Information',
  description: "Description de l'alerte",
  small: true,
};
Petite.storyName = 'Petite';

export const Fermable = Template.bind({});
Fermable.args = {
  title: 'Information',
  description: "Description de l'alerte",
  closable: true,
};
Fermable.storyName = 'Fermable';
