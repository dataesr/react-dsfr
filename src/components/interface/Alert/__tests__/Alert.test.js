import { render, screen } from '@testing-library/react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';
import Alert from '..';

Enzyme.configure({ adapter: new Adapter() });

describe('<Alert />', () => {
  it('should render alert properly', () => {
    render(
      <Alert
        className="custom-class"
        data-testid="alert"
        title="Titre"
        description="Description"
      />,
    );
    const group = screen.getByTestId('alert');
    expect(group).toMatchSnapshot();
  });

  it('should hide alert on close', () => {
    const mockClick = jest.fn();

    const alert = shallow(
      <Alert
        data-testid="alert"
        title="Titre"
        closable
        onClose={mockClick}
      />,
    );

    expect(alert.find('.fr-alert').exists()).toBeTruthy();

    alert.find('.fr-alert > button').simulate('click');

    expect(alert.find('.fr-alert').exists()).toBeFalsy();
    expect(mockClick).toHaveBeenCalled();
  });
});
