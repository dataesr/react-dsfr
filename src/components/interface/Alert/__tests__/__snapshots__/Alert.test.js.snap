// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should render alert properly 1`] = `
<div
  class="fr-alert fr-alert--info custom-class"
  data-testid="alert"
>
  <p
    class="fr-alert__title"
  >
    Titre
  </p>
  <p>
    Description
  </p>
</div>
`;
